scipy/sparse/csgraph/tests/test_shortest_path.py

```diff
@@ -198,6 +199,68 @@ def test_shortest_path_min_only_random(n):
             p = pred[p]
 
 
+# ensure dijkstra does not crash or run into an infinite loop
+@pytest.mark.parametrize('n, t',
+                         ((3, 300),
+                          (10, 100),
+                          (100, 10),
+                          (500, 1)))
+def test_dijkstra_random(n, t):
+    np.random.seed(1234)
+    for testcase_index in range(t):
+        data = scipy.sparse.rand(n, n, density=2.0 / n, format='lil',
+                                 dtype=np.float64)
+        data.setdiag(np.zeros(n, dtype=np.bool_))
+        dijkstra(data, directed=True, return_predecessors=True)
```

Comment:
here's a draft version of the test, with a fixed matrix (no searching), that hangs on `main` and passes on this branch, after doing some brute force searching for a small 10x10 reproducer per CJs suggestion (I also modernized the rng infra while I was there):

```diff
--- a/scipy/sparse/csgraph/tests/test_shortest_path.py
+++ b/scipy/sparse/csgraph/tests/test_shortest_path.py
@@ -200,18 +200,16 @@ def test_dijkstra_min_only_random(n):
 
 
 # ensure dijkstra does not crash or run into an infinite loop
-@pytest.mark.parametrize('n, t',
-                         ((3, 300),
-                          (10, 100),
-                          (100, 10),
-                          (500, 1)))
-def test_dijkstra_random(n, t):
-    np.random.seed(1234)
-    for testcase_index in range(t):
-        data = scipy.sparse.rand(n, n, density=2.0 / n, format='lil',
-                                 dtype=np.float64)
-        data.setdiag(np.zeros(n, dtype=np.bool_))
-        dijkstra(data, directed=True, return_predecessors=True)
+def test_dijkstra_random():
+    n = 10
+    # this random seed was carefully chosen in a search
+    # for reproducers of gh-17782
+    rng = np.random.default_rng(3443)
+    data = scipy.sparse.rand(n, n, density=2.0 / n, format='lil',
+                             dtype=np.float64,
+                             random_state=rng)
+    data.setdiag(np.zeros(n, dtype=np.bool_))
+    dijkstra(data, directed=True, return_predecessors=True)
```